### PR TITLE
Changes to the Kyber code to get lax typechecking

### PR DESCRIPTION
**Pull request by xvzcf**

### CODE REVIEW

File: src/kem/kyber/arithmetic.rs [hunk 1]
```diff
@@ -2,39 +2,60 @@ use super::constants::{COEFFICIENTS_IN_RING_ELEMENT, FIELD_MODULUS};
...
+const INVERSE_OF_MODULUS_MOD_R: u32 = 62209; // FIELD_MODULUS^{-1} mod MONTGOMERY_R
 
-#[cfg_attr(hax, hax_lib_macros::requires(value > -FIELD_MODULUS * MONTGOMERY_R / 2 && value < FIELD_MODULUS * MONTGOMERY_R / 2))]
-#[cfg_attr(hax, hax_lib_macros::ensures(|result| result > -FIELD_MODULUS && result < FIELD_MODULUS))]
+#[cfg_attr(hax, hax_lib_macros::requires(value >= -FIELD_MODULUS * MONTGOMERY_R && value <= FIELD_MODULUS * MONTGOMERY_R))]
+#[cfg_attr(hax, hax_lib_macros::ensures(|result| result >= -(3 * FIELD_MODULUS) / 2 && result <= (3 * FIELD_MODULUS) / 2))]
 pub(crate) fn montgomery_reduce(value: KyberFieldElement) -> KyberFieldElement {
-    let k = (value & (MONTGOMERY_R - 1)) * INVERSE_OF_MODULUS_MOD_R;
-    let k = (k & (MONTGOMERY_R - 1)) as i16;
+    // This forces hax to extract code for MONTGOMERY_R before it extracts code
```
Comment (franziskuskiefer) [inline src/kem/kyber/arithmetic.rs:32]: Please add an issue to anything that should get changed again, like this.
Maybe just file one issue and link everything to it.

File: src/kem/kyber/arithmetic.rs [hunk 2]
```diff
@@ -2,39 +2,60 @@ use super::constants::{COEFFICIENTS_IN_RING_ELEMENT, FIELD_MODULUS};
...
+    // for this function. It is needed due to:
+    // https://github.com/hacspec/hax/issues/332
+    let _ = MONTGOMERY_R;
 
-    let c = (i32::from(k) * FIELD_MODULUS) >> MONTGOMERY_SHIFT;
+    hax_lib::debug_assert!(
```
Comment (franziskuskiefer) [inline src/kem/kyber/arithmetic.rs:37]: Please file an issue for how to get rid of this dependency for regular builds.

File: src/kem/kyber/sampling.rs [hunk 3]
```diff
@@ -9,34 +9,39 @@ pub fn sample_from_uniform_distribution<const SEED_SIZE: usize>(
 ) -> (KyberPolynomialRingElement, Option<Error>) {
     let mut sampled_coefficients: usize = 0;
     let mut out: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
+    let mut done = false;
```
Comment (franziskuskiefer) [inline src/kem/kyber/sampling.rs:12]: Add an issue to fix this again.

File: src/kem/kyber/types.rs [hunk 4]
```diff
@@ -1,7 +1,7 @@
 macro_rules! impl_generic_struct {
     ($name:ident) => {
         pub struct $name<const SIZE: usize> {
-            value: [u8; SIZE],
+            pub(crate) value: [u8; SIZE],
```
Comment (franziskuskiefer) [inline src/kem/kyber/types.rs:4]: Let's not make it so visible ...

```suggestion
            pub(in super) value: [u8; SIZE],
```

File: Cargo.toml [hunk 5]
```diff
@@ -43,6 +43,7 @@ rand = { version = "0.8" }
 log = { version = "0.4", optional = true }
 # WASM API
 wasm-bindgen = { version = "0.2.87", optional = true }
+hax-lib = { git = "https://github.com/hacspec/hax/", branch = "karthik/core-for-kyber" }
```
Comment (franziskuskiefer) [inline Cargo.toml:46]: This will need to be updated before we can merge this.
Has this all been in https://github.com/hacspec/hax/pull/345?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Can you extract the F* code again? It looks quite different with the current version of hax.
Also, lax typechecking doesn't go through for me. Looks like something is missing/has changed somewhere.
```
* Error 19 at Libcrux.Kem.Kyber.Arithmetic.fst(28,2-28,62):
  - Could not prove post-condition
```

File: Cargo.toml (hunk 5 above)
Comment (W95Psp) [reply Cargo.toml:46]: Yes, `karthik/core-for-kyber` and `main` have now the same `proofs-libs`

File: Cargo.toml [hunk 6]
```diff
@@ -43,7 +43,7 @@ rand = { version = "0.8" }
 log = { version = "0.4", optional = true }
 # WASM API
 wasm-bindgen = { version = "0.2.87", optional = true }
-hax-lib = { git = "https://github.com/hacspec/hax/", branch = "karthik/core-for-kyber" }
+hax-lib = { git = "https://github.com/hacspec/hax/", branch = "main" }
```
Comment (franziskuskiefer) [inline Cargo.toml:46]: We shouldn't need a branch here?

**Review by franziskuskiefer — APPROVED** [review]
Thanks, lgtm with that one nit

File: Cargo.toml (hunk 6 above)
Comment (xvzcf) [reply Cargo.toml:46]: I get the following error when I remove the branch:

```
error: failed to download `hax-lib v0.1.0-pre.1 (https://github.com/hacspec/hax/#f31b319e)`

Caused by:
  unable to get packages from source

Caused by:
  failed to find hax-lib v0.1.0-pre.1 (https://github.com/hacspec/hax/#f31b319e) in path source
note: this is an unexpected cargo internal error
note: we would appreciate a bug report: https://github.com/rust-lang/cargo/issues/
note: cargo 1.76.0-nightly (6790a5127 2023-11-10)
```

**Pull request merged**